File: mochi/runtime/agent_run_packages.py
"""Shared package builders for agent-run exports and snapshots."""

from __future__ import annotations

from collections import Counter
from datetime import UTC, datetime
from typing import Any, Mapping

ATTEMPT_BUNDLE_MANIFEST_VERSION = "mochi.agent_run.attempt_bundle.v1"
DATASET_PACKAGE_MANIFEST_VERSION = "mochi.agent_run.dataset_package.v1"


def build_attempt_bundle(
    run: Mapping[str, Any],
    *,
    attempt_id: str | None,
    selected_scope: str,
) -> dict[str, Any]:
    """Build an attempt-scoped replay/evaluator package."""

    artifacts = _scoped_artifacts(run, attempt_id)
    events = _scoped_events(run, attempt_id)
    role_outputs = _extract_role_outputs(events)
    evaluation_events = [
        event for event in events if str(event.get("type") or "") == "evaluation"
    ]
    dataset_records = [
        dict(artifact.get("metadata", {}).get("record") or {})
        for artifact in artifacts
        if str(artifact.get("artifact_type") or "") == "dataset_record"
        and isinstance(artifact.get("metadata", {}).get("record"), dict)
    ]
    selected_candidate_id = _selected_candidate_id(run, dataset_records=dataset_records)
    final_answer = _final_answer(run, dataset_records=dataset_records)

    payload = {
        "manifest_version": ATTEMPT_BUNDLE_MANIFEST_VERSION,
        "package_type": "attempt_bundle",
        "exported_at": _now_iso(),
        "run_id": str(run.get("run_id") or run.get("id") or ""),
        "protocol_id": str(run.get("protocol_id") or ""),
        "attempt_id": attempt_id,
        "selected_scope": selected_scope,
        "schedule_attempt": _schedule_attempt(run, attempt_id),
        "artifact_count": len(artifacts),
        "event_count": len(events),
        "role_output_count": len(role_outputs),
        "artifacts": [_serialize_artifact(artifact) for artifact in artifacts],
        "events": [dict(event) for event in events],
        "role_outputs": role_outputs,
        "evaluation_events": [dict(event) for event in evaluation_events],
        "dataset_records": dataset_records,
        "run_summary": _run_summary_payload(run, artifacts, attempt_id),
        "evidence_summary": _artifact_content(artifacts, "evidence_summary"),
        "verification_summary": _artifact_content(artifacts, "verification_summary"),
        "final_selected_candidate": {
            "candidate_id": selected_candidate_id,
            "final_answer": final_answer,
        }
        if selected_candidate_id is not None or final_answer is not None
        else None,
    }
    payload["replay_ready"] = bool(events) and payload["final_selected_candidate"] is not None
    return payload


def build_dataset_package(run: Mapping[str, Any]) -> dict[str, Any]:
    """Build a training-facing dataset package for an entire run."""

    run_id = str(run.get("run_id") or run.get("id") or "")
    protocol_id = str(run.get("protocol_id") or "")
    recent_attempts = _record_array(_schedule_payload(run).get("recent_attempts"))
    attempt_lookup = {
        attempt_id: attempt
        for attempt in recent_attempts
        if (attempt_id := _string(attempt.get("attempt_id"))) is not None
    }
    grouped: dict[str, list[dict[str, Any]]] = {}
    grouped_attempt_records: dict[str, list[dict[str, Any]]] = {}

    for artifact in _artifacts(run):
        if str(artifact.get("artifact_type") or "") != "dataset_record":
            continue
        record = artifact.get("metadata", {}).get("record")
        if not isinstance(record, dict):
            continue
        attempt_id = _artifact_attempt_id(artifact) or "unscoped"
        governed_record = _govern_dataset_record(
            run,
            attempt_id=None if attempt_id == "unscoped" else attempt_id,
            artifact=artifact,
            record=record,
        )
        grouped.setdefault(attempt_id, []).append(governed_record)
        grouped_attempt_records.setdefault(attempt_id, []).append(governed_record)

    attempts: list[dict[str, Any]] = []
    all_records: list[dict[str, Any]] = []
    training_ready_records: list[dict[str, Any]] = []
    excluded_reason_counts: Counter[str] = Counter()

    ordered_attempt_ids = [
        attempt_id
        for attempt_id in attempt_lookup
        if attempt_id in grouped_attempt_records
    ]
    for attempt_id in grouped_attempt_records:
        if attempt_id != "unscoped" and attempt_id not in ordered_attempt_ids:
            ordered_attempt_ids.append(attempt_id)
    if "unscoped" in grouped_attempt_records:
        ordered_attempt_ids.append("unscoped")

    for grouped_attempt_id in ordered_attempt_ids:
        records = grouped_attempt_records.get(grouped_attempt_id, [])
        if not records:
            continue
        training_ready = [record for record in records if bool(record.get("training_ready"))]
        excluded = [record for record in records if not bool(record.get("training_ready"))]
        for record in excluded:
            for reason in _string_array(record.get("exclusion_reasons")):
                excluded_reason_counts[reason] += 1
        attempts.append(
            {
                "attempt_id": None if grouped_attempt_id == "unscoped" else grouped_attempt_id,
                "schedule_attempt": None
                if grouped_attempt_id == "unscoped"
                else attempt_lookup.get(grouped_attempt_id),
                "dataset_record_count": len(records),
                "training_ready_count": len(training_ready),
                "excluded_record_count": len(excluded),
                "dataset_records": records,
            }
        )
        all_records.extend(records)
        training_ready_records.extend(training_ready)

    excluded_records_summary = {
        "excluded_count": len([record for record in all_records if not bool(record.get("training_ready"))]),
        "reasons": [
            {"reason": reason, "count": count}
            for reason, count in sorted(excluded_reason_counts.items())
        ],
    }

    return {
        "manifest_version": DATASET_PACKAGE_MANIFEST_VERSION,
        "package_type": "dataset_package",
        "exported_at": _now_iso(),
        "run_id": run_id,
        "protocol_id": protocol_id,
        "attempt_count": len(attempts),
        "dataset_record_count": len(all_records),
        "training_ready_count": len(training_ready_records),
        "excluded_record_count": excluded_records_summary["excluded_count"],
        "attempts": attempts,
        "all_records": all_records,
        "training_ready_records": training_ready_records,
        "excluded_records_summary": excluded_records_summary,
    }


def attempt_id_exists(run: Mapping[str, Any], attempt_id: str) -> bool:
    """Check whether an attempt id is present in the run payload."""

    if _schedule_attempt(run, attempt_id) is not None:
        return True
    if any(_artifact_attempt_id(artifact) == attempt_id for artifact in _artifacts(run)):
        return True
    return any(_event_attempt_id(event) == attempt_id for event in _events(run))


def summarize_package_materialization(
    attempt_bundle: Mapping[str, Any],
    dataset_package: Mapping[str, Any],
    *,
    attempt_id: str | None,
    package_error: str | None = None,
) -> dict[str, Any]:
    """Summarize package health for schedule metadata."""

    package_attempt_id = attempt_id or "unscoped"
    attempt_entry = next(
        (
            entry
            for entry in _record_array(dataset_package.get("attempts"))
            if _string(entry.get("attempt_id")) == attempt_id
            or (attempt_id is None and entry.get("attempt_id") is None)
        ),
        None,
    )
    package_ready = package_error is None
    return {
        "attempt_id": package_attempt_id,
        "artifact_count": int(attempt_bundle.get("artifact_count") or 0) + 2,
        "dataset_record_count": int(
            (attempt_entry or {}).get("dataset_record_count") or len(_record_array(attempt_bundle.get("dataset_records")))
        ),
        "training_ready_count": int((attempt_entry or {}).get("training_ready_count") or 0),
        "package_ready": package_ready,
        "last_package_error": package_error,
        "last_package_completed_at": _now_iso() if package_ready else None,
    }


def _govern_dataset_record(
    run: Mapping[str, Any],
    *,
    attempt_id: str | None,
    artifact: Mapping[str, Any],
    record: Mapping[str, Any],
) -> dict[str, Any]:
    selected_models = _selected_models_by_role(run)
    selected_candidate_id = _selected_candidate_id(run, dataset_records=[record])
    final_answer = _final_answer(run, dataset_records=[record])
    verification_summary = _artifact_content(
        _scoped_artifacts(run, attempt_id),
        "verification_summary",
    )
    verification_status = _verification_status(verification_summary, selected_candidate_id)
    evidence_gate_status = _evidence_gate_status(record, selected_candidate_id)

    exclusion_reasons: list[str] = []
    if not final_answer:
        exclusion_reasons.append("missing_final_answer")
    if verification_status == "failed":
        exclusion_reasons.append("verification_failed")
    if evidence_gate_status == "failed":
        exclusion_reasons.append("evidence_gate_failed")

    return {
        "artifact_id": artifact.get("artifact_id"),
        "title": artifact.get("title"),
        "uri": artifact.get("uri"),
        "run_id": str(run.get("run_id") or run.get("id") or ""),
        "attempt_id": attempt_id,
        "protocol_id": str(run.get("protocol_id") or ""),
        "selected_candidate_id": selected_candidate_id,
        "verification_status": verification_status,
        "evidence_gate_status": evidence_gate_status,
        "teacher_model_id": selected_models.get("teacher"),
        "student_model_id": selected_models.get("student"),
        "judge_model_id": selected_models.get("judge"),
        "verifier_model_id": selected_models.get("verifier"),
        "training_ready": len(exclusion_reasons) == 0,
        "exclusion_reasons": exclusion_reasons,
        "record": dict(record),
    }


def _selected_models_by_role(run: Mapping[str, Any]) -> dict[str, str | None]:
    payload = run.get("selected_models_roles")
    if not isinstance(payload, dict):
        return {}
    by_role = payload.get("by_role")
    if not isinstance(by_role, dict):
        return {}
    return {
        role: _string(model_id)
        for role, model_id in by_role.items()
        if isinstance(role, str)
    }


def _verification_status(
    verification_summary: Mapping[str, Any] | None,
    selected_candidate_id: str | None,
) -> str | None:
    if selected_candidate_id is None or not isinstance(verification_summary, Mapping):
        return None
    for item in _record_array(verification_summary.get("verifications")):
        if _string(item.get("candidate_id")) == selected_candidate_id:
            return _string(item.get("status"))
    if selected_candidate_id in _string_array(verification_summary.get("failed_candidate_ids")):
        return "failed"
    if selected_candidate_id in _string_array(verification_summary.get("verified_candidate_ids")):
        return "verified"
    if selected_candidate_id in _string_array(verification_summary.get("skipped_candidate_ids")):
        return "skipped"
    return None


def _evidence_gate_status(
    record: Mapping[str, Any],
    selected_candidate_id: str | None,
) -> str | None:
    evidence = record.get("evidence")
    if not isinstance(evidence, dict):
        return None
    evaluation = evidence.get("evaluation")
    if not isinstance(evaluation, dict):
        return None
    for score in _record_array(evaluation.get("scores")):
        if selected_candidate_id is not None and _string(score.get("candidate_id")) != selected_candidate_id:
            continue
        gate = score.get("evidence_gate")
        if isinstance(gate, dict):
            return _string(gate.get("status"))
    return None


def _selected_candidate_id(
    run: Mapping[str, Any],
    *,
    dataset_records: list[Mapping[str, Any]],
) -> str | None:
    for record in dataset_records:
        target = record.get("target")
        if isinstance(target, dict):
            selected = _string(target.get("candidate_id"))
            if selected is not None:
                return selected
    summary = run.get("summary")
    if isinstance(summary, dict):
        selected = _string(summary.get("selected_candidate_id"))
        if selected is not None:
            return selected
    return None


def _final_answer(
    run: Mapping[str, Any],
    *,
    dataset_records: list[Mapping[str, Any]],
) -> str | None:
    for record in dataset_records:
        target = record.get("target")
        if isinstance(target, dict):
            answer = _string(target.get("answer"))
            if answer is not None:
                return answer
    summary = run.get("summary")
    if isinstance(summary, dict):
        answer = _string(summary.get("final_answer"))
        if answer is not None:
            return answer
    return None


def _run_summary_payload(
    run: Mapping[str, Any],
    artifacts: list[dict[str, Any]],
    attempt_id: str | None,
) -> dict[str, Any]:
    summary = run.get("summary")
    payload = dict(summary) if isinstance(summary, dict) else {}
    artifact_payload = _artifact_content(artifacts, "run_summary")
    if isinstance(artifact_payload, dict):
        payload["artifact_content"] = artifact_payload
    payload["attempt_id"] = attempt_id
    return payload


def _extract_role_outputs(events: list[dict[str, Any]]) -> list[dict[str, Any]]:
    outputs: list[dict[str, Any]] = []
    for event in events:
        if str(event.get("type") or "") != "role_output":
            continue
        payload = event.get("payload")
        if not isinstance(payload, dict):
            continue
        outputs.append(
            {
                "role_id": _string(payload.get("role_id")) or "unknown",
                "candidate_id": _string(payload.get("candidate_id")),
                "model_id": _string(payload.get("model_id")),
                "round_index": payload.get("round_index")
                if isinstance(payload.get("round_index"), int)
                else None,
                "content": _string(payload.get("content")) or "",
                "timestamp": _string(event.get("timestamp")),
            }
        )
    return outputs


def _serialize_artifact(artifact: Mapping[str, Any]) -> dict[str, Any]:
    return {
        "artifact_id": artifact.get("artifact_id"),
        "artifact_type": artifact.get("artifact_type"),
        "title": artifact.get("title"),
        "uri": artifact.get("uri"),
        "mime_type": artifact.get("mime_type"),
        "size_bytes": artifact.get("size_bytes"),
        "metadata": dict(artifact.get("metadata") or {}),
    }


def _artifact_content(
    artifacts: list[dict[str, Any]],
    artifact_type: str,
) -> dict[str, Any] | None:
    for artifact in artifacts:
        if str(artifact.get("artifact_type") or "") != artifact_type:
            continue
        metadata = artifact.get("metadata")
        if isinstance(metadata, dict) and isinstance(metadata.get("content"), dict):
            return dict(metadata["content"])
    return None


def _scoped_artifacts(run: Mapping[str, Any], attempt_id: str | None) -> list[dict[str, Any]]:
    artifacts = _artifacts(run)
    if attempt_id is None:
        return [artifact for artifact in artifacts if _artifact_attempt_id(artifact) is None]
    return [artifact for artifact in artifacts if _artifact_attempt_id(artifact) == attempt_id]


def _scoped_events(run: Mapping[str, Any], attempt_id: str | None) -> list[dict[str, Any]]:
    events = _events(run)
    if attempt_id is None:
        return [event for event in events if _event_attempt_id(event) is None]
    return [event for event in events if _event_attempt_id(event) == attempt_id]


def _schedule_attempt(run: Mapping[str, Any], attempt_id: str | None) -> dict[str, Any] | None:
    if attempt_id is None:
        return None
    for attempt in _record_array(_schedule_payload(run).get("recent_attempts")):
        if _string(attempt.get("attempt_id")) == attempt_id:
            return attempt
    return None


def _events(run: Mapping[str, Any]) -> list[dict[str, Any]]:
    return [dict(event) for event in _record_array(run.get("events"))]


def _artifacts(run: Mapping[str, Any]) -> list[dict[str, Any]]:
    return [dict(artifact) for artifact in _record_array(run.get("artifacts"))]


def _schedule_payload(run: Mapping[str, Any]) -> dict[str, Any]:
    schedule = run.get("schedule")
    return dict(schedule) if isinstance(schedule, dict) else {}


def _artifact_attempt_id(artifact: Mapping[str, Any]) -> str | None:
    metadata = artifact.get("metadata")
    if not isinstance(metadata, dict):
        return None
    return _string(metadata.get("attempt_id"))


def _event_attempt_id(event: Mapping[str, Any]) -> str | None:
    return _string(event.get("attempt_id"))


def _record_array(value: Any) -> list[dict[str, Any]]:
    if not isinstance(value, list):
        return []
    return [dict(item) for item in value if isinstance(item, dict)]


def _string_array(value: Any) -> list[str]:
    if not isinstance(value, list):
        return []
    return [item for item in value if isinstance(item, str) and item.strip()]


def _string(value: Any) -> str | None:
    if isinstance(value, str):
        stripped = value.strip()
        return stripped or None
    return None


def _now_iso() -> str:
    return datetime.now(UTC).isoformat()

File: web/src/lib/agent-run-packages.ts
import type * as api from '@/lib/api'

export const ATTEMPT_BUNDLE_MANIFEST_VERSION = 'mochi.agent_run.attempt_bundle.v1'
export const DATASET_PACKAGE_MANIFEST_VERSION = 'mochi.agent_run.dataset_package.v1'

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null
}

function getString(value: unknown): string | null {
  return typeof value === 'string' && value.trim().length > 0 ? value : null
}

function getNullableString(value: unknown): string | null {
  return getString(value)
}

function getRecordArray(value: unknown): Array<Record<string, unknown>> {
  return Array.isArray(value) ? value.filter(isRecord) : []
}

function getArtifactAttemptId(artifact: api.AgentRunArtifact): string | null {
  return getString(artifact.metadata.attempt_id)
}

function getEventAttemptId(event: Record<string, unknown>): string | null {
  return getString(event.attempt_id)
}

function getScheduleAttempts(run: api.AgentRunDetail): Array<Record<string, unknown>> {
  return getRecordArray(run.schedule.recent_attempts)
}

function getScheduleAttempt(
  run: api.AgentRunDetail,
  attemptId: string | null
): Record<string, unknown> | null {
  if (!attemptId) {
    return null
  }
  return (
    getScheduleAttempts(run).find((attempt) => getString(attempt.attempt_id) === attemptId) ?? null
  )
}

function scopedArtifacts(run: api.AgentRunDetail, attemptId: string | null, scope: string): api.AgentRunArtifact[] {
  if (scope === 'all') {
    return run.artifacts
  }
  if (!attemptId) {
    return run.artifacts.filter((artifact) => getArtifactAttemptId(artifact) === null)
  }
  return run.artifacts.filter((artifact) => getArtifactAttemptId(artifact) === attemptId)
}

function scopedEvents(
  run: api.AgentRunDetail,
  attemptId: string | null,
  scope: string
): Array<Record<string, unknown>> {
  if (scope === 'all') {
    return run.events
  }
  if (!attemptId) {
    return run.events.filter((event) => getEventAttemptId(event) === null)
  }
  return run.events.filter((event) => getEventAttemptId(event) === attemptId)
}

function selectedCandidateId(
  run: api.AgentRunDetail,
  datasetRecords: Array<Record<string, unknown>>
): string | null {
  for (const record of datasetRecords) {
    const target = isRecord(record.target) ? record.target : null
    const candidateId = getString(target?.candidate_id)
    if (candidateId) {
      return candidateId
    }
  }
  return getString(run.summary.selected_candidate_id)
}

function finalAnswer(
  run: api.AgentRunDetail,
  datasetRecords: Array<Record<string, unknown>>
): string | null {
  for (const record of datasetRecords) {
    const target = isRecord(record.target) ? record.target : null
    const answer = getString(target?.answer)
    if (answer) {
      return answer
    }
  }
  return getString(run.summary.final_answer)
}

function artifactContent(
  artifacts: api.AgentRunArtifact[],
  artifactType: string
): Record<string, unknown> | null {
  const artifact = artifacts.find((item) => item.artifact_type === artifactType) ?? null
  return isRecord(artifact?.metadata.content) ? artifact.metadata.content : null
}

function extractCollectorShardManifests(
  artifacts: api.AgentRunArtifact[]
): Array<Record<string, unknown>> {
  return artifacts
    .filter((artifact) => artifact.artifact_type === 'collector_shard_manifest')
    .map((artifact) => {
      const content = isRecord(artifact.metadata.content) ? artifact.metadata.content : {}
      return {
        ...content,
        artifact_id: artifact.artifact_id,
        title: artifact.title,
        uri: artifact.uri,
      }
    })
}

function filterCollectorShardManifests(
  manifests: Array<Record<string, unknown>>,
  attemptId: string | null
): Array<Record<string, unknown>> {
  return manifests.filter((manifest) => getString(manifest.attempt_id) === attemptId)
}

function buildCollectorProvenanceManifest(
  records: Array<Record<string, unknown>>
): Record<string, unknown> | null {
  const manifestRecords: Array<Record<string, unknown>> = []
  records.forEach((item, index) => {
    const record = isRecord(item.record) ? item.record : item
    const metadata = isRecord(record.metadata) ? record.metadata : null
    const provenance = isRecord(metadata?.collector_provenance) ? metadata.collector_provenance : null
    if (!provenance) {
      return
    }
    manifestRecords.push({
      ...provenance,
      record_index: index + 1,
      artifact_id: getString(item.artifact_id),
      attempt_id: getString(item.attempt_id),
    })
  })
  if (manifestRecords.length === 0) {
    return null
  }

  const countByValue = (key: string) => {
    const counts = new Map<string, number>()
    for (const item of manifestRecords) {
      const value = getString(item[key])
      if (!value) {
        continue
      }
      counts.set(value, (counts.get(value) ?? 0) + 1)
    }
    return Array.from(counts.entries())
      .sort(([left], [right]) => left.localeCompare(right))
      .map(([value, count]) => ({ value, count }))
  }

  return {
    schema_version: '1.0',
    record_count: manifestRecords.length,
    records: manifestRecords,
    adapter_counts: countByValue('adapter_name'),
    policy_disposition_counts: countByValue('policy_disposition'),
    license_counts: countByValue('license'),
  }
}

function selectedModelsByRole(run: api.AgentRunDetail): Record<string, string | null> {
  const payload = isRecord(run.selected_models_roles.by_role) ? run.selected_models_roles.by_role : null
  if (!payload) {
    return {}
  }
  return Object.fromEntries(
    Object.entries(payload)
      .filter(([key, value]) => typeof key === 'string' && typeof value === 'string')
      .map(([key, value]) => [key, String(value)])
  )
}

function verificationStatus(
  verificationSummary: Record<string, unknown> | null,
  candidateId: string | null
): string | null {
  if (!verificationSummary || !candidateId) {
    return null
  }
  for (const item of getRecordArray(verificationSummary.verifications)) {
    if (getString(item.candidate_id) === candidateId) {
      return getString(item.status)
    }
  }
  return null
}

function evidenceGateStatus(record: Record<string, unknown>, candidateId: string | null): string | null {
  const evidence = isRecord(record.evidence) ? record.evidence : null
  const evaluation = isRecord(evidence?.evaluation) ? evidence.evaluation : null
  for (const score of getRecordArray(evaluation?.scores)) {
    if (candidateId && getString(score.candidate_id) !== candidateId) {
      continue
    }
    const gate = isRecord(score.evidence_gate) ? score.evidence_gate : null
    const status = getString(gate?.status)
    if (status) {
      return status
    }
  }
  return null
}

function buildGovernedDatasetRecord(
  run: api.AgentRunDetail,
  artifact: api.AgentRunArtifact,
  record: Record<string, unknown>
): Record<string, unknown> {
  const attemptId = getArtifactAttemptId(artifact)
  const candidateId = selectedCandidateId(run, [record])
  const verificationSummary = artifactContent(
    scopedArtifacts(run, attemptId, attemptId ? attemptId : 'latest'),
    'verification_summary'
  )
  const finalAnswerValue = finalAnswer(run, [record])
  const verification = verificationStatus(verificationSummary, candidateId)
  const evidenceGate = evidenceGateStatus(record, candidateId)
  const exclusionReasons: string[] = []
  if (!finalAnswerValue) {
    exclusionReasons.push('missing_final_answer')
  }
  if (verification === 'failed') {
    exclusionReasons.push('verification_failed')
  }
  if (evidenceGate === 'failed') {
    exclusionReasons.push('evidence_gate_failed')
  }
  const models = selectedModelsByRole(run)

  return {
    artifact_id: artifact.artifact_id,
    title: artifact.title,
    uri: artifact.uri,
    run_id: run.run_id,
    attempt_id: attemptId,
    protocol_id: run.protocol_id,
    selected_candidate_id: candidateId,
    verification_status: verification,
    evidence_gate_status: evidenceGate,
    teacher_model_id: models.teacher ?? null,
    student_model_id: models.student ?? null,
    judge_model_id: models.judge ?? null,
    verifier_model_id: models.verifier ?? null,
    training_ready: exclusionReasons.length === 0,
    exclusion_reasons: exclusionReasons,
    record,
  }
}

export function buildAttemptPackageFallback(
  run: api.AgentRunDetail,
  selectedScope: string,
  attemptId: string | null
): Record<string, unknown> {
  const artifacts = scopedArtifacts(run, attemptId, selectedScope)
  const events = scopedEvents(run, attemptId, selectedScope)
  const roleOutputs = events
    .filter((event) => event.type === 'role_output' && isRecord(event.payload))
    .map((event) => ({
      role_id: getString((event.payload as Record<string, unknown>).role_id) ?? 'unknown',
      candidate_id: getString((event.payload as Record<string, unknown>).candidate_id),
      model_id: getString((event.payload as Record<string, unknown>).model_id),
      round_index:
        typeof (event.payload as Record<string, unknown>).round_index === 'number'
          ? ((event.payload as Record<string, unknown>).round_index as number)
          : null,
      content: getString((event.payload as Record<string, unknown>).content) ?? '',
      timestamp: getString(event.timestamp),
    }))
  const datasetRecords = artifacts
    .filter((artifact) => artifact.artifact_type === 'dataset_record')
    .map((artifact) => artifact.metadata.record)
    .filter((record): record is Record<string, unknown> => isRecord(record))
  const collectorShardManifests = extractCollectorShardManifests(artifacts)
  const candidateId = selectedCandidateId(run, datasetRecords)
  const finalAnswerValue = finalAnswer(run, datasetRecords)

  return {
    manifest_version: ATTEMPT_BUNDLE_MANIFEST_VERSION,
    package_type: 'attempt_bundle',
    exported_at: new Date().toISOString(),
    run_id: run.run_id,
    protocol_id: run.protocol_id,
    attempt_id: attemptId,
    selected_scope: selectedScope,
    schedule_attempt: getScheduleAttempt(run, attemptId),
    artifact_count: artifacts.length,
    event_count: events.length,
    role_output_count: roleOutputs.length,
    replay_ready: events.length > 0 && (candidateId !== null || finalAnswerValue !== null),
    artifacts: artifacts.map((artifact) => ({
      artifact_id: artifact.artifact_id,
      artifact_type: artifact.artifact_type,
      title: artifact.title,
      uri: artifact.uri,
      mime_type: artifact.mime_type,
      size_bytes: artifact.size_bytes,
      metadata: artifact.metadata,
    })),
    events,
    role_outputs: roleOutputs,
    evaluation_events: events.filter((event) => event.type === 'evaluation'),
    collector_shard_manifests: collectorShardManifests,
    collector_provenance_manifest: buildCollectorProvenanceManifest(datasetRecords),
    dataset_records: datasetRecords,
    run_summary: {
      ...run.summary,
      attempt_id: attemptId,
      artifact_content: artifactContent(artifacts, 'run_summary'),
    },
    evidence_summary: artifactContent(artifacts, 'evidence_summary'),
    verification_summary: artifactContent(artifacts, 'verification_summary'),
    final_selected_candidate:
      candidateId !== null || finalAnswerValue !== null
        ? {
            candidate_id: candidateId,
            final_answer: finalAnswerValue,
          }
        : null,
  }
}

export function buildDatasetPackageFallback(run: api.AgentRunDetail): Record<string, unknown> {
  const datasetArtifacts = run.artifacts.filter((artifact) => artifact.artifact_type === 'dataset_record')
  const collectorShardManifests = extractCollectorShardManifests(run.artifacts)
  const grouped = new Map<string, Array<Record<string, unknown>>>()
  const scheduleAttempts = getScheduleAttempts(run)

  for (const artifact of datasetArtifacts) {
    const record = isRecord(artifact.metadata.record) ? artifact.metadata.record : null
    if (!record) {
      continue
    }
    const attemptId = getArtifactAttemptId(artifact) ?? 'unscoped'
    const current = grouped.get(attemptId) ?? []
    current.push(buildGovernedDatasetRecord(run, artifact, record))
    grouped.set(attemptId, current)
  }

  const attempts: Array<Record<string, unknown>> = []
  const allRecords: Array<Record<string, unknown>> = []
  const trainingReadyRecords: Array<Record<string, unknown>> = []
  const excludedReasons = new Map<string, number>()

  const orderedAttemptIds = scheduleAttempts
    .map((attempt) => getString(attempt.attempt_id))
    .filter((attemptId): attemptId is string => Boolean(attemptId))

  for (const attemptId of grouped.keys()) {
    if (attemptId !== 'unscoped' && !orderedAttemptIds.includes(attemptId)) {
      orderedAttemptIds.push(attemptId)
    }
  }
  if (grouped.has('unscoped')) {
    orderedAttemptIds.push('unscoped')
  }

  for (const attemptId of orderedAttemptIds) {
    const records = grouped.get(attemptId) ?? []
    if (records.length === 0) {
      continue
    }
    const trainingReady = records.filter((record) => Boolean(record.training_ready))
    const excluded = records.filter((record) => !record.training_ready)
    for (const record of excluded) {
      const reasons = Array.isArray(record.exclusion_reasons) ? record.exclusion_reasons : []
      for (const reason of reasons) {
        if (typeof reason !== 'string' || reason.length === 0) {
          continue
        }
        excludedReasons.set(reason, (excludedReasons.get(reason) ?? 0) + 1)
      }
    }
    attempts.push({
      attempt_id: attemptId === 'unscoped' ? null : attemptId,
      schedule_attempt:
        attemptId === 'unscoped'
          ? null
          : scheduleAttempts.find((attempt) => getString(attempt.attempt_id) === attemptId) ?? null,
      dataset_record_count: records.length,
      training_ready_count: trainingReady.length,
      excluded_record_count: excluded.length,
      collector_shard_manifests: extractCollectorShardManifests(
        scopedArtifacts(run, attemptId === 'unscoped' ? null : attemptId, attemptId)
      ),
      collector_provenance_manifest: buildCollectorProvenanceManifest(records),
      dataset_records: records,
    })
    allRecords.push(...records)
    trainingReadyRecords.push(...trainingReady)
  }

  return {
    manifest_version: DATASET_PACKAGE_MANIFEST_VERSION,
    package_type: 'dataset_package',
    exported_at: new Date().toISOString(),
    run_id: run.run_id,
    protocol_id: run.protocol_id,
    attempt_count: attempts.length,
    dataset_record_count: allRecords.length,
    training_ready_count: trainingReadyRecords.length,
    excluded_record_count: allRecords.length - trainingReadyRecords.length,
    collector_shard_manifests: collectorShardManifests,
    collector_provenance_manifest: buildCollectorProvenanceManifest(allRecords),
    attempts,
    all_records: allRecords,
    training_ready_records: trainingReadyRecords,
    excluded_records_summary: {
      excluded_count: allRecords.length - trainingReadyRecords.length,
      reasons: Array.from(excludedReasons.entries()).map(([reason, count]) => ({ reason, count })),
    },
  }
}

export function buildTrainingReadyOnlyDatasetPackage(
  datasetPackage: Record<string, unknown>
): Record<string, unknown> {
  const allCollectorShardManifests = getRecordArray(datasetPackage.collector_shard_manifests)
  const attempts: Array<Record<string, unknown>> = getRecordArray(datasetPackage.attempts).map(
    (attempt) => {
      const datasetRecords = getRecordArray(attempt.dataset_records).filter(
        (record) => record.training_ready === true
      )
      const attemptId = getNullableString(attempt.attempt_id)
      return {
      ...attempt,
      collector_shard_manifests: filterCollectorShardManifests(
        allCollectorShardManifests,
        attemptId
      ),
      collector_provenance_manifest: buildCollectorProvenanceManifest(datasetRecords),
      dataset_record_count: datasetRecords.length,
      training_ready_count: datasetRecords.length,
        excluded_record_count: 0,
        dataset_records: datasetRecords,
      }
    }
  )
  const filteredAttempts = attempts.filter(
    (attempt) => typeof attempt.dataset_record_count === 'number' && attempt.dataset_record_count > 0
  )
  const trainingReadyRecords = getRecordArray(datasetPackage.training_ready_records)
  const allowedAttemptIds = new Set(
    filteredAttempts.map((attempt) => getNullableString(attempt.attempt_id))
  )
  return {
    ...datasetPackage,
    dataset_record_count: trainingReadyRecords.length,
    training_ready_count: trainingReadyRecords.length,
    excluded_record_count: 0,
    collector_shard_manifests: allCollectorShardManifests.filter((manifest) =>
      allowedAttemptIds.has(getNullableString(manifest.attempt_id))
    ),
    collector_provenance_manifest: buildCollectorProvenanceManifest(trainingReadyRecords),
    attempts: filteredAttempts,
    all_records: trainingReadyRecords,
    training_ready_records: trainingReadyRecords,
    excluded_records_summary: {
      excluded_count: 0,
      reasons: [],
    },
  }
}
